R/pmcmc_nested_parameters.R

```diff
@@ -0,0 +1,488 @@
+                        names(private$parameters))
+      base[idx_fixed] <- fixed
+      base_transform <- private$transform
+
+      if (is.null(self$populations())) {
+        proposal <- private$proposal_kernel[idx_vary, idx_vary, drop = FALSE]
+      } else {
+        proposal <- array(
+          apply(private$proposal_kernel, 3, "[", idx_vary,
+                                idx_vary, drop = FALSE),
+          dim = c(length(idx_vary), length(idx_vary),
+                  length(self$populations())))
+      }
+
+      transform <- function(p) {
+          base_transform(set_into(base, idx_vary, p))
+      }
+
+      pmcmc_parameters$new(private$parameters[idx_vary], proposal, transform,
+        populations = self$populations())
+    }
+  ))
+
+
+clean_parameters <- function(parameters) {
+  assert_is(parameters, "list")
+  if (length(parameters) == 0) {
+    stop("At least one parameter is required")
+  }
+
+  type <- c("pmcmc_parameter", "pmcmc_varied_parameter")
+  ok <- vlapply(parameters, inherits, type)
+  if (!all(ok)) {
+    stop(sprintf("Expected all elements of '...' to be '%s' objects", type))
+  }
+
+  varied <- vlapply(parameters, inherits, "pmcmc_varied_parameter")
+  if (!any(varied)) {
+    stop("No varied parameters, use `pmcmc_parameters` instead.")
+  } else if (all(varied)) {
+    fixed_parameters <- NULL
+  } else {
+    fixed_parameters <- parameters[!varied]
+  }
+  varied_parameters <- parameters[varied]
+
+  # check fixed and varied names separately
+  if (!is.null(fixed_parameters)) {
+    fix_nms <- vcapply(fixed_parameters, "[[", "name", USE.NAMES = FALSE)
+  } else {
+    fix_nms <- c()
+  }
+
+  # only need to get the name of the first population as internally asserted
+  var_nms <- vcapply(varied_parameters, function(x) x[[1]]$name[1],
+    USE.NAMES = FALSE)
+  nms <- c(fix_nms, var_nms)
+  dups <- nms[duplicated(nms)]
+  if (length(dups) > 0L) {
+    stop("Duplicate parameter names: ",
+         paste(squote(unique(dups)), collapse = ", "))
+  }
+
+  if (!is.null(fixed_parameters)) {
+    if (!is.null(names(fixed_parameters)) &&
+      !identical(fix_nms, names(fixed_parameters))) {
+      stop("Fixed parameters are named, but the names do not match
+      parameters.")
+    }
+    names(fixed_parameters) <- fix_nms
+  }
+
+  if (!is.null(names(varied_parameters)) &&
+    !identical(var_nms, names(varied_parameters))) {
+    stop("Varied parameters are named, but the names do not match parameters.")
+  }
+  names(varied_parameters) <- var_nms
+
+  populations <- names(varied_parameters[[1]])
+  ok <- vlapply(varied_parameters,
+    function(x) identical(names(x), populations))
+  if (!all(ok)) {
+    stop("Populations and ordering of varied parameters should be identical.")
+  }
+
+  list(fixed = fixed_parameters, varied = varied_parameters, pop = populations,
+    names = list(fixed = fix_nms, varied = var_nms))
+}
+
+clean_proposals <- function(proposal_fixed, proposal_varied, populations,
```

Comment:
I think the code below could be clarified if you lift things like `length(params$varied)` into constants at the top of the functions as you use them several times in complicated if statements